### ISSUE TITLE
On-demand Magnet Integration

### ISSUE THREAD

**Issue by wd30130** [Magport/Magnet#4]
We should provide the framework and system in Magnet for ordering the core of relay chain, the program can refer to https://github.com/paritytech/polkadot-sdk/issues/1487, which is almost practical.  But we should make some changes and simplifications. 

1. Core buyer and block author need to establish a relationship, i.e., the buyer of the core is responsible for producing the block, if the buyer can not produce the block, other nodes can assist to produce, so as to comply with the responsibility relationship. 
Specific changes are, in the aura block author selection, judgment if the chain is parathread, then first determine whether the Core buyer now also has the right to produce block, if he has, then select him to produce block, if not, then according to the normal aura selection，select other block author to produce block.

2. Simplify the program of https://github.com/paritytech/polkadot-sdk/issues/1487 , remove the Core purchase proof mechanism and challenge mechanism.
https://github.com/paritytech/polkadot-sdk/issues/1487 need Core purchase proof mechanism and challenge mechanism ( to solve the problem of purchasing a Core, but not making a block on that Core). This part can be removed with the implementation of (1) above. Since the buyer of the Core is responsible for making the block, and others are only assisting, it is the buyer's responsibility if the block are not made. If a Core is purchased and no block are produced on it, the Core purchaser has to bear the loss, and the parathread system does not compensate for it, so the parathread system does not need to record and handle this situation.

By changing and simplifying the above 2 points, the system is simpler and more efficient.

**Comment by eskimor** [comment]
Also, maybe interesting previous iterations of the design: https://github.com/paritytech/cumulus/issues/2154

(Which also aimed at making block orderer and block producer related) In any case, keep it coming. I agree that the proposed design is rather complex, but it is not because I like it complex, it was the simplest I could come up with that fits the threat model. So, really happy to hear ideas for making it simpler.

**Comment by wd30130** [comment]
Yes,  every block should  provide core purchase proof,  our concrete realization used core purchase proof .

**Comment by wd30130** [comment]
Although, the buyer of the core is the first responsibility for producing the block,   nevertheless will make system 
 simpler and more efficient

**Comment by wd30130** [comment]
closed by #19

**Comment by Aideepakchaudhary** [comment]
@wd30130 I wanted to implement similar kind of thing in our Parachain where I can get a on-demand core at certain period of time(let's say 5PM) everyday so I need this operation should be automatic. 
I saw your PR and it look promising what I'm looking for Can you please elaborate a bit how this is achieving the automation and also when the chain doesn't have the core(let's say 12PM), Is collator are running at that time? Can you please give a overview how things are happening end- 2?

**Comment by sulijia** [comment]
@Aideepakchaudhary If you want to get an on-demand core at a certain time every day, you can call place_order_allow_death or place_order_keep_alive on relaychian through a scheduled task.
To achieve automation, one is to execute it at a fixed time every day, and the other is to trigger it according to conditions. We choose the latter. We will determine whether there is a transaction in the transaction pool to trigger the purchase of the core.

**Comment by Aideepakchaudhary** [comment]
@sulijia Thank you for the response. Can you please little elaborate about the scheduled task. It will be very helpful.

In your chain, If there is no transaction in the transaction pool that means the chain is also not running and as soon as a new transaction come your chain start producing block Is that correct? but my question is, If chain is not running then how you accept the new transactions and send it to the relay chain?

**Comment by sulijia** [comment]
@Aideepakchaudhary If there is no transaction in the transaction pool that means the chain is also not running and as soon as a new transaction come your chain start producing block Is that correct? 
Yes.
We construct a transaction through the crate provided by subxt (https://github.com/paritytech/subxt), which contains the call of place_order_allow_death or place_order_keep_alive, and then sign it with the private key in the chain keystore, and then submit the transaction through the RPC interface provided by relaychain.
This part of the logic runs in the node part, not in the runtime, and does not require chain running

**Comment by Aideepakchaudhary** [comment]
@sulijia I tried with to scheduler the call through scheduler in rococo extrisincs like this:
![image](https://github.com/user-attachments/assets/48191fc3-8396-4ea6-a30f-4042f132fd66)

but I'm getting these event:
<img width="698" alt="Screenshot 2024-08-02 at 10 53 50 PM" src="https://github.com/user-attachments/assets/5a2aa3da-e796-43a3-8514-b1e0518b022f">

can you please suggest something how I can schedule this call? Or it’s not possible?

**Comment by sulijia** [comment]
@Aideepakchaudhary BadOrigin is because you are using sudo, you should use it like this
![Uploading 20240803-085722.jpg…]()

In addition, you need to ensure that there are cores allocated to the Coretime Pool. You can achieve this in this way.
![Uploading 20240803-090306.jpg…]()

**Comment by sulijia** [comment]
@Aideepakchaudhary BadOrigin is because you are using sudo, you should use it like this
![20240803-085722](https://github.com/user-attachments/assets/6bc6ac9f-04b7-4d13-a7cf-08cd1ba7cc6d)

In addition, you need to ensure that there are cores allocated to the Coretime Pool. You can achieve this in this way.
![20240803-090306](https://github.com/user-attachments/assets/8f817937-ff61-408d-be40-78c70b32a734)
